dhall-lsp-server/src/Backend/Dhall/Diagnostics.hs

```diff
@@ -188,3 +192,40 @@ errorFancyLength :: Text.Megaparsec.ShowErrorComponent e => Text.Megaparsec.Erro
 errorFancyLength = \case
   Text.Megaparsec.ErrorCustom a -> Text.Megaparsec.errorComponentLen a
   _             -> 1
+
+-- sanitise range to exclude surrounding whitespace
+-- makes sure that the resulting range is well-formed
+sanitiseRange :: Range -> Text -> Range
+sanitiseRange (Range l r) text = Range l (max l r')
+  where r' = trimEndPosition r text
+
+-- Variants of T.lines and T.unlines that are inverses of one another.
+lines' :: Text -> [Text]
```

Comment:
Perhaps restrict the type to `Text -> NonEmpty Text`.  My understanding is that `Data.Text.split` always returns a non-empty list even if the type does not convey that information.  You could implement it as:

```haskell
lines' :: Text -> NonEmpty Text
lines' text =
    case Data.Text.split (== '\n') text of
        []     -> "" :| []
        l : ls -> l  :| ls
```

The main reason I suggest this is so that functions like `offsetToPosition` don't need to handle the case where the list is empty since you've upstreamed the logic to handle that into the `lines'` function

Reply:
Thanks for spotting this! I changed the definition to the one you proposed.